### PR TITLE
refactor: add sys params to load partial templates

### PR DESCRIPTION
**Pull request by macintoshpie**
#### What does this PR accomplish?
- Adds building sys params to:
  - load model instance partial templates
  - component definitions partial templates (if coupling includes a load)
  - connect statements partial templates (if coupling includes a load)

Also includes changes from #419 which was merged into this PR's branch. TLDR is we moved some sys params around so that we can also drop them into other couplings.

Also includes changes from #418 which was merged into this PR's branch. TLDR we used the sys params from the district_system in the network 2 pipe <> Heating Plant coupling file

#### How should this be manually tested?
- Edit one of the templates to use `sys_params`, run a test which uses the template, verify it was templated into the resulting model.

#### What are the relevant tickets?
First step into solving #401 

#### Screenshots (if appropriate)
Example templating into a coupling file (⚠️ building sys params have moved since this screenshot, they're now at `sys_params.building`)
Screen Shot 2021-08-04 at 10 29 22 AM

Screen Shot 2021-08-04 at 10 28 56 AM

Example templating into an instance file (⚠️ building sys params have moved since this screenshot, they're now at `sys_params.building`)
Screen Shot 2021-08-04 at 10 30 10 AM

Screen Shot 2021-08-04 at 10 29 47 AM

Example templating with district_system params
Screen Shot 2021-08-13 at 4 47 56 PM

Screen Shot 2021-08-13 at 4 47 38 PM

### CODE REVIEW

File: docs/developer_resources.rst [hunk 1]
```diff
@@ -141,23 +141,25 @@ Model Definition
...
-This is the template which defines connect statements to be inserted into the equation section.
+This is the template which defines connect statements to be inserted into the equation section. More specifically, these are the partial template files at model_connectors/couplings/templates/<coupling name>/ConnectStatements.mopt. These templates have access to:
 
-- :code:`globals`
-- :code:`coupling`: just like with the component definitions template, you should use the coupling.id to avoid variable name collisions.
-- :code:`graph`
+- :code:`globals`: same as Coupling Component Definitions context
+- :code:`coupling`: same as Coupling Component Definitions context. Just like with the component definitions template, you should use the coupling.id to avoid variable name collisions.
+- :code:`graph`: same as Coupling Component Definitions context
+- :code:`sys_params`: same as Coupling Component Definitions context
```
Comment (nllong) [inline docs/developer_resources.rst:162]: nice! 🏆

File: docs/developer_resources.rst [hunk 2]
```diff
@@ -168,9 +170,10 @@ This template is used to declare a model instance.
 - :code:`graph`
 - :code:`couplings`: contains each coupling the model is associated with. For example, if our ETS was coupled to a load and network, couplings would look like :code:`{ load_couplings: [<load coupling>], network_couplings: [<network coupling>] }`. This can be used to access coupling and model ids.
 - :code:`model`: contains info about the model instance, including :code:`modelica_type` and :code:`id`. These should be used to define the model, for example :code:`{{ model.modelica_type }} {{ model.id }}(...)`
+- :code:`sys_params`: if the model is a load, this object will be included as well -- if the model is not a load this object will not be present. It contains the building-specific system parameters pulled from the system parameters JSON file.
```
Comment (nllong) [inline docs/developer_resources.rst:173]:
```suggestion
- :code:`sys_params`: this object will be included if the model type is a load; reflexively, this data object will not be present if the model type is not a load. It contains the building-specific system parameters pulled from the system parameters JSON file.
```

File: geojson_modelica_translator/model_connectors/couplings/coupling.py [hunk 3]
```diff
@@ -188,6 +188,22 @@ def render_templates(self, template_params):
             'connect_statements_template_path': connect_template_path,
         }
 
+    def get_load(self):
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/couplings/coupling.py:191]: nice function! 🧮

**Review by nllong — APPROVED** [review]
baller

**Comment by macintoshpie** [comment]
@vtnate are you going to tackle adding the district_system part of sys params to the couplings? Happy to help or do it myself. i.e. point 3 in comment here:
https://github.com/urbanopt/geojson-modelica-translator/issues/401#issuecomment-896323096
> refactor district.py to pass the district_system block to coupling mopt files

I suppose we could also do that in your draft PR (#418  ) as well if you'd rather.

**Comment by macintoshpie** [comment]
@nllong @vtnate please review the most recent changes

File: geojson_modelica_translator/model_connectors/districts/district.py [hunk 4]
```diff
@@ -117,7 +116,10 @@ def to_modelica(self):
                 'delHeaWatTemBui': 'delHeaWatTemBui',
                 'delHeaWatTemDis': 'delHeaWatTemDis',
             },
-            'graph': self._coupling_graph
+            'graph': self._coupling_graph,
+            'sys_params': {
+                'district_system': self.system_parameters.get_param('$.district_system')
```
Comment (macintoshpie) [inline geojson_modelica_translator/model_connectors/districts/district.py:121]: kind of assumes all sys params will have district_system now, but I think that's ok?

**Review by vtnate — APPROVED** [review]
Lot of moving parts to this little change to have it behave the way we expected it always should! LGTM.

File: geojson_modelica_translator/model_connectors/couplings/templates/Network2Pipe_HeatingPlant/ComponentDefinitions.mopt [hunk 5]
```diff
@@ -7,6 +7,6 @@
   // TODO: This should not be here, it is entirely plant specific and should be moved elsewhere
   // but since it requires a connect statement we must put it here for now...
   Modelica.Blocks.Sources.RealExpression TDisSetHeaWat_{{ coupling.id }}(
-    each y=55+273.15)
+    each y=273.15+{{ sys_params.district_system.default.central_heating_plant_parameters.temp_setpoint_hhw }})
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/couplings/templates/Network2Pipe_HeatingPlant/ComponentDefinitions.mopt:10]: 🔥

File: geojson_modelica_translator/model_connectors/districts/district.py (hunk 4 above)
Comment (nllong) [reply geojson_modelica_translator/model_connectors/districts/district.py:121]: this seems reasonable.

File: geojson_modelica_translator/model_connectors/districts/district.py [hunk 6]
```diff
@@ -114,14 +116,23 @@ def to_modelica(self):
...
         for coupling in self._coupling_graph.couplings:
             template_context = {
                 'diagram': diagram.to_dict(coupling.id, is_coupling=True),
             }
             template_context.update(**common_template_params)
+
+            coupling_load = coupling.get_load()
+            if coupling_load is not None:
+                building_sys_params = self.system_parameters.get_param_by_building_id(coupling_load.building_id, '$')
+                template_context['sys_params']['building'] = building_sys_params
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/districts/district.py:134]: nice, i was wondering how you were constraining to the selected building.

**Review by nllong — APPROVED** [review]
Looks amazing!

@mingzhe37  -- note that this may conflict with your updates. Make sure to favor these changes!

**Pull request merged**